staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go

```diff
@@ -230,8 +231,7 @@ func (wc *watchChan) run(initialEventsEndBookmarkRequired, forceInitialEvents bo
 	go wc.startWatching(watchClosedCh, initialEventsEndBookmarkRequired, forceInitialEvents)
 
 	var resultChanWG sync.WaitGroup
-	resultChanWG.Add(1)
-	go wc.processEvent(&resultChanWG)
```

Comment:
previously we had 1 goroutine here per watch, now we have up to 12 (schedule + collect + 10 processFunc)? I don't have a good sense for goroutine counts / watch counts and whether that multiplier is reasonable

Reply:
Described in commit message. In my testing, that mostly included running SVM stress test, I have noticed that no matter the run, there is a noticeable correlation of maximum and the average processing time of an event. If average was 10ms, the max was around 100ms, if average was 6ms the max was around 60ms. Based on that I proposed that in cases that if there is a spike in latency we could run 10 concurrent processing to maintain same latency.

Reply:
I can provide more data based on analysing different concurrency.

Reply:
I agree with that, but goroutines have other impact too (e.g. creating goroutine allocates memory, etc.).
I guess what Jordan is asking for is thinking about this change in a wider context.

Reply:
I cannot give answers for context I don't have. If you can provide direct asks about what testing/benchmarks you would like to see to ensure safety then I can provide that.  The flaw with serial processing of watch events was exactly due to lack of testing.

Reply:
So here are a bit deeper thoughts from me:
- in general all watches (except from those specifying explicitly RV="") are served from watchcache, so the overhead (in terms of resource consumption) of "etcd3 watches" is not that critical given that context
- here we're replacing 1 goroutine with 2 goroutines, so that's for sure fine
- so what may have a non-negligible impact is:
(a) additional channel [currently of size 10] per watch
(b) the fact that each call to transform is now spawning a gorotuine

For (a) I'm not worried, given the first point above (and given in watchcache we have a buffer of size 10 for every single watch too].

For (b) this should be fine given the optimizations were put into golang itself in this area. But I would like to see the impact of it measure as difference in resource consumption (CPU and RAM) of kube-apiserver between:
(a) this PR
(b) this PR change so that we don't spawn a goroutine for transformation [which obviously doesn't address the problem then]